PERF: Speed-up common case of loadtxt()ing non-hex floats.

anntzer opened the pull request:
`python runtests.py --bench bench_io` reports a ~5-10% perf gain.
(It's not huge, but it's a very low-hanging fruit.)

See also https://github.com/numpy/numpy/issues/19541.

Example benchmark runs:

Before:
```
[ 78.95%] ··· bench_io.LoadtxtCSVComments.time_comment_loadtxt_csv                                                ok
[ 78.95%] ··· =========== ============
               num_lines              
              ----------- ------------
                   10      51.7±0.1μs 
                  100       336±3μs   
                 10000     31.7±0.8ms 
                 100000     329±2ms   
              =========== ============

[ 81.58%] ··· bench_io.LoadtxtCSVDateTime.time_loadtxt_csv_datetime                                               ok
[ 81.58%] ··· =========== =============
               num_lines               
              ----------- -------------
                   20       92.0±0.4μs 
                  200        677±3μs   
                  2000     6.44±0.03ms 
                 20000      65.5±0.2ms 
              =========== =============

[ 84.21%] ··· bench_io.LoadtxtCSVSkipRows.time_skiprows_csv                                                       ok
[ 84.21%] ··· ========== =========
               skiprows           
              ---------- ---------
                  0       402±3ms 
                 500      403±3ms 
                10000     362±1ms 
              ========== =========

[ 86.84%] ··· bench_io.LoadtxtCSVStructured.time_loadtxt_csv_struct_dtype                                  255±0.4ms
[ 89.47%] ··· bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv                                                   ok
[ 89.47%] ··· ============ ============ =========== ============= ===========
              --                               num_lines                     
              ------------ --------------------------------------------------
                 dtype          10          100         10000        100000  
              ============ ============ =========== ============= ===========
                float32     53.1±0.7μs    334±2μs     31.0±0.1ms    327±4ms  
                float64     52.9±0.4μs    337±2μs     31.3±0.2ms    324±1ms  
                 int32      53.5±0.8μs    328±1μs     30.6±0.3ms    315±2ms  
                 int64      54.6±0.3μs    360±1μs    33.6±0.08ms    351±1ms  
               complex128   57.4±0.3μs    380±1μs     35.1±0.1ms    365±1ms  
                  str       55.4±0.4μs    347±1μs     32.2±0.1ms    330±2ms  
                 object     51.0±0.3μs   318±0.9μs    29.7±0.1ms   308±0.6ms 
              ============ ============ =========== ============= ===========
```

After:
```
[ 78.95%] ··· bench_io.LoadtxtCSVComments.time_comment_loadtxt_csv                                                ok
[ 78.95%] ··· =========== ============
               num_lines              
              ----------- ------------
                   10      50.6±0.3μs 
                  100       314±3μs   
                 10000     29.6±0.1ms 
                 100000     310±2ms   
              =========== ============

[ 81.58%] ··· bench_io.LoadtxtCSVDateTime.time_loadtxt_csv_datetime                                               ok
[ 81.58%] ··· =========== =============
               num_lines               
              ----------- -------------
                   20       90.1±0.4μs 
                  200        662±4μs   
                  2000     6.37±0.06ms 
                 20000      63.9±0.9ms 
              =========== =============

[ 84.21%] ··· bench_io.LoadtxtCSVSkipRows.time_skiprows_csv                                                       ok
[ 84.21%] ··· ========== =========
               skiprows           
              ---------- ---------
                  0       376±4ms 
                 500      375±6ms 
                10000     335±6ms 
              ========== =========

[ 86.84%] ··· bench_io.LoadtxtCSVStructured.time_loadtxt_csv_struct_dtype                                  250±0.6ms
[ 89.47%] ··· bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv                                                   ok
[ 89.47%] ··· ============ ============ ========= ============= ===========
              --                              num_lines                    
              ------------ ------------------------------------------------
                 dtype          10         100        10000        100000  
              ============ ============ ========= ============= ===========
                float32     50.0±0.1μs   309±1μs     29.3±2ms     305±9ms  
                float64     50.6±0.4μs   315±5μs   28.9±0.07ms   302±0.9ms 
                 int32      51.3±0.4μs   319±2μs   29.9±0.05ms   308±0.5ms 
                 int64      53.9±0.3μs   350±1μs   32.8±0.09ms    341±1ms  
               complex128   56.9±0.3μs   370±3μs    35.0±0.2ms    359±1ms  
                  str       54.7±0.5μs   339±1μs    31.9±0.2ms    326±1ms  
                 object     50.5±0.2μs   326±2μs    30.5±0.7ms    305±2ms  
              ============ ============ ========= ============= ===========
```

anntzer committed:
PERF: Speed-up common case of loadtxt()ing non-hex floats.

`python runtests.py --bench bench_io` reports a ~5-10% perf gain.

rossbar reviewed (commented):
I see the same ~5-10% speedups: output of `python runtests.py --bench-compare main bench_io` run on this branch:

```
       before           after         ratio
     [2c1a34da]       [c8a2314a]
     <main>           <loadtxtfloat>
-        32.8±1ms       31.2±0.1ms     0.95  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('float32', 10000)
-         344±9ms          327±1ms     0.95  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('float32', 100000)
-         967±9μs          916±2μs     0.95  bench_io.LoadtxtReadUint64Integers.time_read_uint64_neg_values(550)
-      50.4±0.3μs       47.7±0.5μs     0.95  bench_io.LoadtxtCSVComments.time_comment_loadtxt_csv(10)
-     1.75±0.04ms      1.66±0.01ms     0.95  bench_io.LoadtxtReadUint64Integers.time_read_uint64_neg_values(1000)
-         344±8ms          325±2ms     0.95  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('float64', 100000)
-       334±0.4μs        315±0.4μs     0.94  bench_io.LoadtxtCSVComments.time_comment_loadtxt_csv(100)
-         331±1ms          311±2ms     0.94  bench_io.LoadtxtCSVComments.time_comment_loadtxt_csv(100000)
-         976±5μs          915±7μs     0.94  bench_io.LoadtxtReadUint64Integers.time_read_uint64(550)
-      32.0±0.1ms       30.0±0.2ms     0.94  bench_io.LoadtxtCSVComments.time_comment_loadtxt_csv(10000)
-      17.3±0.2ms      16.1±0.06ms     0.93  bench_io.LoadtxtReadUint64Integers.time_read_uint64(10000)
-       405±0.9ms        376±0.7ms     0.93  bench_io.LoadtxtCSVSkipRows.time_skiprows_csv(0)
-         407±2ms          374±2ms     0.92  bench_io.LoadtxtCSVSkipRows.time_skiprows_csv(500)
-        346±10μs        316±0.7μs     0.91  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('float32', 100)
-      96.8±0.8μs       87.2±0.8μs     0.90  bench_io.LoadtxtCSVDateTime.time_loadtxt_csv_datetime(20)
-      18.2±0.6ms       16.2±0.1ms     0.89  bench_io.LoadtxtReadUint64Integers.time_read_uint64_neg_values(10000)
-        383±20ms          341±1ms     0.89  bench_io.LoadtxtCSVSkipRows.time_skiprows_csv(10000)
-      20.0±0.4ms      17.7±0.07ms     0.89  bench_io.LoadtxtUseColsCSV.time_loadtxt_usecols_csv([1, 3, 5, 7])
```

One concern though is the loss of specifity when switching to the try/except. For example there are some subtle behavior changes:

#### Before
```python
>>> l = ['a', 'b', 'c']
>>> np.loadtxt(l)  # float is the default dtype
Traceback (most recent call last)
   ...
ValueError: could not convert string to float: 'a'
```
#### After
```python
>>> l = ['a', 'b', 'c']
>>> np.loadtxt(l)
array([10., 11., 12.])
```

This would also mean that the default exception message for failed parsing of floats would change:

#### Before
```python
>>> l = ['1.23.4']  # 
>>> np.loadtxt(l)
Traceback (most recent call last)
   ...
ValueError: could not convert string to float: '1.23.4'
```
#### After
```python
>>> np.loadtxt(l)
Traceback (most recent call last)
   ...
ValueError: invalid hexadecimal floating-point string
```

I think both of these changes are undesirable and IMO would be blockers for going the try/except route (at least without some additional logic/exception catching, which would probably reduce/eliminate the performance gain).

anntzer commented:
I added some more logic in the except branch to handle these cases and restore the original error message; because they are normally not exercised, the performance speedup remains present:
```
       before           after         ratio
     [f25905b5]       [f3f2e72a]
     <main>           <loadtxtfloat>
+     1.32±0.01μs      1.47±0.02μs     1.12  bench_io.Copy.time_memcpy('int16')
-      33.1±0.4ms       31.5±0.2ms     0.95  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('float32', 10000)
-         345±2ms          328±3ms     0.95  bench_io.LoadtxtCSVComments.time_comment_loadtxt_csv(100000)
-       354±0.7μs          336±3μs     0.95  bench_io.LoadtxtCSVComments.time_comment_loadtxt_csv(100)
-      9.66±0.2ms      9.16±0.05ms     0.95  bench_io.LoadtxtUseColsCSV.time_loadtxt_usecols_csv(2)
-      57.6±0.3μs       54.5±0.3μs     0.95  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('float32', 10)
-      33.4±0.2ms       31.6±0.1ms     0.95  bench_io.LoadtxtCSVComments.time_comment_loadtxt_csv(10000)
-         346±2ms          327±2ms     0.94  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('float64', 100000)
-         427±1ms        403±0.7ms     0.94  bench_io.LoadtxtCSVSkipRows.time_skiprows_csv(0)
-      17.9±0.1ms       16.9±0.1ms     0.94  bench_io.LoadtxtReadUint64Integers.time_read_uint64_neg_values(10000)
-         428±2ms          402±1ms     0.94  bench_io.LoadtxtCSVSkipRows.time_skiprows_csv(500)
-      19.7±0.1ms       18.4±0.2ms     0.94  bench_io.LoadtxtUseColsCSV.time_loadtxt_usecols_csv([1, 3, 5, 7])
-     1.81±0.02ms      1.69±0.01ms     0.93  bench_io.LoadtxtReadUint64Integers.time_read_uint64(1000)
-      17.9±0.1ms       16.7±0.1ms     0.93  bench_io.LoadtxtReadUint64Integers.time_read_uint64(10000)
-         388±2ms          360±2ms     0.93  bench_io.LoadtxtCSVSkipRows.time_skiprows_csv(10000)
-        16.2±1μs      14.7±0.01μs     0.90  bench_io.Copy.time_strided_copy('complex64')
```
(not sure what's happening with bench_io.Copy though)

rossbar commented on numpy/lib/npyio.py:
```diff
@@ -733,10 +733,15 @@ def _getconv(dtype):
-        if '0x' in x:
-            return float.fromhex(x)
-        return float(x)
+        try:
+            return float(x)
+        except ValueError as exc:
+            try:
+                if '0x' in x:
+                    return float.fromhex(x)
+            except ValueError:
+                pass
+            raise exc
```
It might be good to add a comment here to explain why the nested try's & ifs.

anntzer replied:
Done.

rossbar reviewed (commented):
This LGTM - do you think it'd be worth to convert my previous comments into two tests? If you have no objection I'd be happy to push that up.

anntzer commented:
If you can write the tests, please do so and push them :)

rossbar committed:
TST: Add tests to check fromhex not called unintentionally.

Adds regression tests to check that the logic surrounding when
the default floatconv applies the fromhex conversion does not
change.

eric-wieser commented on numpy/lib/npyio.py:
```diff
@@ -733,10 +733,15 @@ def _getconv(dtype):
+                    return float.fromhex(x)
+            except ValueError:
+                pass
+            raise exc  # Raise the original exception, which makes more sense.
```
```suggestion
            raise  # Raise the original exception, which makes more sense.
```

anntzer replied:
It wasn't clear to me that the original exception would indeed be reraised (not the new one), but I checked that that's the case :)

eric-wieser commented on numpy/lib/npyio.py:
```diff
@@ -733,10 +733,15 @@ def _getconv(dtype):
+            return float(x)  # The fastest path.
+        except ValueError as exc:
+            try:
+                if '0x' in x:  # Don't accidentally convert "a" ("0xa") to 10.
```
I think it would be clearer with this if outside the try

anntzer replied:
sure

eric-wieser commented on numpy/lib/npyio.py:
```diff
@@ -733,10 +733,15 @@ def _getconv(dtype):
-        return float(x)
+        try:
+            return float(x)  # The fastest path.
+        except ValueError as exc:
```
```suggestion
        except ValueError:
```

eric-wieser commented on numpy/lib/npyio.py:
```diff
@@ -733,10 +733,15 @@ def _getconv(dtype):
     """ Find the correct dtype converter. Adapted from matplotlib """
 
     def floatconv(x):
-        x.lower()
```
Well this line was stupid!

rossbar commented:
Great, thanks @anntzer & @eric-wieser !
